File: WorkcraftCore/src/main/java/org/workcraft/dom/visual/connections/BezierGui.java
package org.workcraft.dom.visual.connections;

import java.awt.Shape;
import java.awt.geom.CubicCurve2D;
import java.awt.geom.Point2D;
import java.awt.geom.Rectangle2D;

import org.workcraft.util.Geometry;
import org.workcraft.util.Geometry.CurveSplitResult;

public class BezierGui {
	
	public static ParametricCurve makeCurve(VisualConnectionProperties connectionInfo, VisualConnectionContext context, RelativePoint cp1, RelativePoint cp2) {
		CubicCurve2D curve2D = new CubicCurve2D.Double();
		Point2D c1 = context.component1().getCenter();
		Point2D c2 = context.component2().getCenter();
		Point2D absoluteCp1 = cp1.toSpace(c1, c2);
		Point2D absoluteCp2 = cp2.toSpace(c1, c2);
		curve2D.setCurve(c1, absoluteCp1, absoluteCp2, c2);
		return new Curve(curve2D);
	}
	
	private final static class Curve implements ParametricCurve {
		
		@Override
		public Shape getShape(double tStart, double tEnd) {
			CurveSplitResult firstSplit = Geometry.splitCubicCurve(fullCurve2D, tStart);
			CurveSplitResult secondSplit = Geometry.splitCubicCurve(firstSplit.curve2, (tEnd-tStart)/(1-tStart));
			return secondSplit.curve1;
		}
		
		public Curve(CubicCurve2D fullCurve2D) {
			this.fullCurve2D = fullCurve2D;
		}
		
		private final CubicCurve2D fullCurve2D;
		
		@Override
		public double getNearestPointT(Point2D pt) {
			// FIXME: should be done using some proper algorithm
			Double nearest = 0.0;
			double nearestDist = Double.MAX_VALUE;
			
			for (double t=0.01; t<=1.0; t+=0.01) {
				Point2D samplePoint = Geometry.getPointOnCubicCurve(fullCurve2D, t);
				double distance = pt.distance(samplePoint);
				if (distance < nearestDist)	{
					nearestDist = distance;
					nearest = t;
				}
			}
			
			return nearest;
		}

		@Override
		public Point2D getPointOnCurve(double t) {
			return Geometry.getPointOnCubicCurve(fullCurve2D, t);
		}
		@Override
		public Point2D getDerivativeAt(double t) {
			return Geometry.getDerivativeOfCubicCurve(fullCurve2D, t);
		}

		@Override
		public Point2D getSecondDerivativeAt(double t) {
			return Geometry.getSecondDerivativeOfCubicCurve(fullCurve2D, t);
		}
		
		@Override
		public Rectangle2D getBoundingBox() {
			Rectangle2D boundingBox = fullCurve2D.getBounds2D();
			boundingBox.add(boundingBox.getMinX()-VisualConnection.HIT_THRESHOLD, boundingBox.getMinY()-VisualConnection.HIT_THRESHOLD);
			boundingBox.add(boundingBox.getMinX()-VisualConnection.HIT_THRESHOLD, boundingBox.getMaxY()+VisualConnection.HIT_THRESHOLD);
			boundingBox.add(boundingBox.getMaxX()+VisualConnection.HIT_THRESHOLD, boundingBox.getMinY()-VisualConnection.HIT_THRESHOLD);
			boundingBox.add(boundingBox.getMaxX()+VisualConnection.HIT_THRESHOLD, boundingBox.getMaxY()+VisualConnection.HIT_THRESHOLD);
			return boundingBox;
		}
	}
}

File: ScalaGraphEditorUtil/src/main/scala/org/workcraft/dom/visual/connections/BezierGui.scala
package org.workcraft.dom.visual.connections

import java.awt.Shape
import java.awt.geom.CubicCurve2D
import java.awt.geom.Point2D
import java.awt.geom.Rectangle2D
import org.workcraft.graphics.Java2DDecoration._
import org.workcraft.graphics.VisualCurveProperties
import org.workcraft.graphics.Geometry.complexMultiply
import org.workcraft.graphics.Geometry.complexInverse
import org.workcraft.graphics.Java2DDecoration._
import org.workcraft.graphics.ParametricCurve

/**
 * A connection control point coordinate, in a coordinate system where the first connected component is at (0, 0) and the second one is at (1, 0)
 */
case class RelativePoint(point: Point2D.Double) {
  def toUserSpace(p1: Point2D.Double, p2: Point2D.Double) = complexMultiply(point, p2 - p1) + p1
}

object RelativePoint {
  val oneThird = new RelativePoint(new Point2D.Double(1.0 / 3.0, 0.0))
  val twoThirds = new RelativePoint(new Point2D.Double(2.0 / 3.0, 0.0))

  def fromUserSpace(p1: Point2D.Double, p2: Point2D.Double, p: Point2D.Double) = complexInverse(p2 - p1).map(complexMultiply(p - p1, _))
}

object BezierGui {

  def makeCurve(context: VisualConnectionContext, cp1: RelativePoint, cp2: RelativePoint): ParametricCurve = {
    val curve2D = new CubicCurve2D.Double
    val c1 = context.c1.center
    val c2 = context.c2.center
    val absoluteCp1 = cp1.toUserSpace(c1, c2)
    val absoluteCp2 = cp2.toUserSpace(c1, c2)
    curve2D.setCurve(c1, absoluteCp1, absoluteCp2, c2)
    new Curve(curve2D)
  }

  private class Curve(fullCurve2D: CubicCurve2D.Double) extends ParametricCurve {

    override def shape(tStart: Double, tEnd: Double): Shape = {
      val firstSplit = fullCurve2D.split(tStart)
      val secondSplit = firstSplit._2.split((tEnd - tStart) / (1 - tStart))
      secondSplit._1
    }

    override def nearestPointT(pt: Point2D.Double): Double = {
      // FIXME: should be done using some proper algorithm
      // DONT FIXME: it ain't broken :)
      var nearest = 0.0
      var nearestDist = java.lang.Double.MAX_VALUE

      Range.Double(0.01, 1.0, 0.01).foreach(t => {
        val samplePoint = fullCurve2D.getPointOnCurve(t)
        val distance = pt.distance(samplePoint)
        if (distance < nearestDist) {
          nearestDist = distance
          nearest = t
        }
      })

      nearest
    }

    override def pointOnCurve(t: Double): Point2D.Double = fullCurve2D.getPointOnCurve(t)

    override def derivativeAt(t: Double): Point2D.Double = fullCurve2D.getDerivative(t)

    override def secondDerivativeAt(t: Double): Point2D.Double = fullCurve2D.getSecondDerivative(t)

    override def boundingBox: Rectangle2D.Double = {
      val boundingBox = fullCurve2D.bounds
      boundingBox.add(boundingBox.getMinX - VisualConnectionGui.HitThreshold, boundingBox.getMinY - VisualConnectionGui.HitThreshold)
      boundingBox.add(boundingBox.getMinX - VisualConnectionGui.HitThreshold, boundingBox.getMaxY + VisualConnectionGui.HitThreshold)
      boundingBox.add(boundingBox.getMaxX + VisualConnectionGui.HitThreshold, boundingBox.getMinY - VisualConnectionGui.HitThreshold)
      boundingBox.add(boundingBox.getMaxX + VisualConnectionGui.HitThreshold, boundingBox.getMaxY + VisualConnectionGui.HitThreshold)
      boundingBox
    }
  }
}
